cmd/skopeo/sync.go

```diff
@@ -531,19 +619,23 @@ func (opts *syncOptions) run(args []string, stdout io.Writer) (retErr error) {
 	if len(opts.source) == 0 {
 		return errors.New("A source transport must be specified")
 	}
-	if !contains(opts.source, []string{docker.Transport.Name(), directory.Transport.Name(), "yaml"}) {
+	if !contains(opts.source, []string{docker.Transport.Name(), directory.Transport.Name(), "yaml", oci.Transport.Name()}) {
 		return errors.Errorf("%q is not a valid source transport", opts.source)
 	}
 
 	if len(opts.destination) == 0 {
 		return errors.New("A destination transport must be specified")
 	}
-	if !contains(opts.destination, []string{docker.Transport.Name(), directory.Transport.Name()}) {
+	if !contains(opts.destination, []string{docker.Transport.Name(), directory.Transport.Name(), oci.Transport.Name()}) {
 		return errors.Errorf("%q is not a valid destination transport", opts.destination)
 	}
 
-	if opts.source == opts.destination && opts.source == directory.Transport.Name() {
-		return errors.New("sync from 'dir' to 'dir' not implemented, consider using rsync instead")
+	if opts.source == opts.destination && contains(opts.source, []string{directory.Transport.Name(), oci.Transport.Name()}) {
+		return errors.Errorf("sync from '%s' to '%s' not implemented, consider using rsync instead", opts.source, opts.destination)
+	}
+
+	if opts.preserveDigests && opts.destination == oci.Transport.Name() {
+		return errors.New("oci DESTINATION transports not support --preserve-digests options")
```

Comment:
Note to self: All of this looks increasingly clumsy; take a deeper look.

Reply:
(To clarify, I meant the existing code structure — it feels like a collection of special cases disconnected from the actual implementation, and can get out of sync.)

Reply:
Yeah, let’s split this up, to remove the transport list duplication, and so that `imagesToCopy` and `destinationReference` don’t grow very large.

Turn `imagesToCopy` into a
```go
var sourceImageListers map[string]func(…)… = {
   docker.Transport.Name(): listDockerSourceImages,
   …
}
```
(`imagesToCopy` has almost no shared code, so that can be just carved up)

and similarly `destinationReference` should turn into
```go
var destinationReferenceConstructors map[string]func(…)… = {
   docker.Transport.Name(): createDockerDestinationReference,
   …
}
```

Then these `contains` checks turn into
```go
sourceImageLister, ok := sourceImageListers[opts.source]
if !ok { fail }
```

---

That feels like a necessary but arguably unrelated cleanup — I can work on that in a separate PR, if you’d like.